### PR TITLE
Deprecate torch.lu_solve

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73878
* #73877
* #74046
* __->__ #73806
* #73804
* #73803
* #72935

**BC-breaking note**:

This PR deprecates `torch.lu_solve` in favor of `torch.linalg.lu_solve_factor`.
A upgrade guide is added to the documentation for `torch.lu_solve`.

Note this PR DOES NOT remove `torch.lu_solve`.

### CODE REVIEW

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -1979,10 +1979,26 @@ TORCH_IMPL_FUNC(linalg_lu_solve_out)(const Tensor& LU,
 }
 
 Tensor lu_solve(const Tensor& self, const Tensor& LU_data, const Tensor& LU_pivots) {
+  TORCH_WARN_ONCE(
+    "torch.lu_solve is deprecated in favor of torch.linalg.lu_solve",
+    "and will be removed in a future PyTorch release.\n",
+    "Note that torch.linalg.solve has its arguments reversed.\n",
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1985]:
```suggestion
    "Note that torch.linalg.lu_solve has its arguments reversed.\n",
```

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -1979,10 +1979,26 @@ TORCH_IMPL_FUNC(linalg_lu_solve_out)(const Tensor& LU,
...
+    "X = torch.linalg.lu_solve(LU, pivots, B)"
+  );
   return at::linalg_lu_solve(LU_data, LU_pivots, self);
 }
 
 Tensor& lu_solve_out(const Tensor& self, const Tensor& LU_data, const Tensor& LU_pivots, Tensor& result) {
+  TORCH_WARN_ONCE(
+    "torch.lu_solve is deprecated in favor of torch.linalg.lu_solve",
+    "and will be removed in a future PyTorch release.\n",
+    "Note that torch.linalg.solve has its arguments reversed.\n",
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1997]:
```suggestion
    "Note that torch.linalg.lu_solve has its arguments reversed.\n",
```

**Review by IvanYashchuk — APPROVED** [review]

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:1985]: Aaah, yes, the copy-paste

**Review by nikitaved — APPROVED** [review]

**Review by mruberry — APPROVED** [review]
Cool! Stamped!

**Comment by Lezcano** [comment]
Relanding in https://github.com/pytorch/pytorch/pull/77637

**Pull request closed without merging**